gossip/services/tcp_receiver.cc

```diff
@@ -19,54 +22,101 @@ static logger_id_t logger_id;
+
+  rw_lock_handle_wrlock(&service_->state->node->neighbors_lock);
 
   neighbor_t* neighbor = neighbors_find_by_endpoint_values(
-      service_->state->node->neighbors, host, port, PROTOCOL_TCP);
+      service_->state->node->neighbors, remote_host_.c_str(), remote_port_,
+      PROTOCOL_TCP);
 
   if (neighbor == NULL) {
     log_info(logger_id,
-             "Connection denied with non-tethered neighbor tcp://%s:%d\n", host,
-             port);
+             "Connection denied with non-tethered neighbor tcp://%s:%d\n",
+             remote_host_.c_str(), remote_port_);
     rw_lock_handle_unlock(&service_->state->node->neighbors_lock);
     return;
   }
 
   log_info(logger_id,
-           "Connection accepted with tethered neighbor tcp://%s:%d\n", host,
-           port);
+           "Connection accepted with tethered neighbor tcp://%s:%d\n",
+           remote_host_.c_str(), remote_port_);
+
+  // Opening connection with neighbor
+
+  boost::asio::ip::tcp::endpoint neighbor_endpoint(
+      boost::asio::ip::address::from_string(remote_host_), remote_port_);
+  auto neighbor_socket =
+      new boost::asio::ip::tcp::socket(socket_.get_executor().context());
+  neighbor_socket->connect(neighbor_endpoint);
+
+  // Sending listening port to neighbor
+
+  std::stringstream stream;
+  stream << std::setfill('0') << std::setw(PORT_SIZE) << remote_port_;
+  boost::asio::write(*neighbor_socket, boost::asio::buffer(stream.str()),
+                     error);
+
+  neighbor->endpoint.opaque_inetaddr = neighbor_socket;
+
   rw_lock_handle_unlock(&service_->state->node->neighbors_lock);
 
-  receive();
-}
+  iota_packet_t packet;
+
+  for (;;) {
+    std::array<char, PACKET_SIZE + CRC_SIZE> tcp_packet;
+
+    boost::asio::read(socket_, boost::asio::buffer(tcp_packet), error);
+
+    if (error) {
+      log_warning(logger_id,
+                  "Reading from tethered node tcp://%s:%d failed: %s\n",
+                  remote_host_.c_str(), remote_port_, error.message().c_str());
+      break;
+    }
+
+    // Computing CRC
+
+    boost::crc_32_type result;
+    result.process_bytes(&tcp_packet[0], PACKET_SIZE);
+    stream.clear();
+    stream.str(std::string());
+    stream << std::setfill('0') << std::setw(CRC_SIZE) << std::hex
+           << result.checksum();
 
-void TcpConnection::receive() {
-  auto self(shared_from_this());
-  boost::asio::async_read(
-      socket_, boost::asio::buffer(packet_.content, PACKET_SIZE),
-      [this, self](boost::system::error_code ec, std::size_t length) {
-        if (!ec && length == PACKET_SIZE) {
-          iota_packet_set_endpoint(
-              &packet_, socket_.remote_endpoint().address().to_string().c_str(),
-              socket_.remote_endpoint().port(), PROTOCOL_TCP);
-          processor_on_next(service_->processor, packet_);
-        }
-        receive();
-      });
+    // Checking CRC
+
+    if (stream.str() == std::string(&tcp_packet[0] + PACKET_SIZE, CRC_SIZE)) {
```

Comment:
after computing the crc of PACKET_SIZE bytes, we only need to memcmp(computed, &tcp_packet[0] + PACKET_SIZE, CRC_SIZE) == 0

std::string constructor copies